diff --git a/ggml/src/ggml-cpu/ggml-cpu.cpp b/ggml/src/ggml-cpu/ggml-cpu.cpp
index e16cdc9d4..b8d1c41e5 100644
--- a/ggml/src/ggml-cpu/ggml-cpu.cpp
+++ b/ggml/src/ggml-cpu/ggml-cpu.cpp
@@ -632,6 +632,18 @@ static void * ggml_backend_cpu_get_proc_address(ggml_backend_reg_t reg, const ch
     if (strcmp(name, "ggml_backend_cpu_is_numa") == 0) {
         return (void *)ggml_is_numa;
     }
+    if (strcmp(name, "ggml_graph_compute_with_ctx") == 0) {
+        return (void *)ggml_graph_compute_with_ctx;
+    }
+    if (strcmp(name, "ggml_set_f32") == 0) {
+        return (void *)ggml_set_f32;
+    }
+    if (strcmp(name, "ggml_set_f32_1d") == 0) {
+        return (void *)ggml_set_f32_1d;
+    }
+    if (strcmp(name, "ggml_backend_is_cpu") == 0) {
+        return (void *)ggml_backend_is_cpu;
+    }
 
     // threadpool - TODO:  move to ggml-base
     if (strcmp(name, "ggml_threadpool_new") == 0) {
