diff --git a/ggml/include/ggml-metal.h b/ggml/include/ggml-metal.h
index 669c1f84..f5d1892c 100644
--- a/ggml/include/ggml-metal.h
+++ b/ggml/include/ggml-metal.h
@@ -61,6 +61,8 @@ GGML_BACKEND_API void ggml_backend_metal_capture_next_compute(ggml_backend_t bac
 
 GGML_BACKEND_API ggml_backend_reg_t ggml_backend_metal_reg(void);
 
+GGML_BACKEND_API void ggml_backend_metal_get_device_memory(ggml_backend_t backend, size_t * free_mem, size_t * total_mem);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/ggml/src/ggml-metal/ggml-metal.m b/ggml/src/ggml-metal/ggml-metal.m
index 093ae900..034ebfdb 100644
--- a/ggml/src/ggml-metal/ggml-metal.m
+++ b/ggml/src/ggml-metal/ggml-metal.m
@@ -2875,7 +2875,7 @@ static void ggml_metal_encode_node(
             } break;
         case GGML_OP_RMS_NORM:
             {
-                GGML_ASSERT(ne00 % 4 == 0);
+                // GGML_ASSERT(ne00 % 4 == 0);
                 GGML_ASSERT(ggml_is_contiguous_1(src0));
 
                 float eps;
@@ -4729,4 +4729,8 @@ ggml_backend_reg_t ggml_backend_metal_reg(void) {
     return &g_ggml_backend_metal_reg;
 }
 
+void ggml_backend_metal_get_device_memory(ggml_backend_t backend, size_t *free, size_t *total) {
+    ggml_backend_metal_device_get_memory(backend->device, free, total);
+}
+
 GGML_BACKEND_DL_IMPL(ggml_backend_metal_reg)
