diff --git a/ggml/src/ggml-backend.cpp b/ggml/src/ggml-backend.cpp
index 273075f4..51e6f39a 100644
--- a/ggml/src/ggml-backend.cpp
+++ b/ggml/src/ggml-backend.cpp
@@ -609,7 +609,7 @@ static bool ggml_is_view_op(enum ggml_op op) {
 // scheduler
 
 #ifndef GGML_SCHED_MAX_BACKENDS
-#define GGML_SCHED_MAX_BACKENDS 16
+#define GGML_SCHED_MAX_BACKENDS 128
 #endif
 
 #ifndef GGML_SCHED_MAX_SPLIT_INPUTS
diff --git a/src/llama.cpp b/src/llama.cpp
index 4a4e9149..1cb577bb 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -9522,7 +9522,7 @@ struct llama_sampler_chain_params llama_sampler_chain_default_params() {
 }
 
 size_t llama_max_devices(void) {
-    return 16;
+    return 128;
 }
 
 bool llama_supports_mmap(void) {
