diff --git a/include/llama.h b/include/llama.h
index 0295a51f..14987887 100644
--- a/include/llama.h
+++ b/include/llama.h
@@ -928,6 +928,7 @@ extern "C" {
     LLAMA_API llama_token llama_token_sep(const struct llama_model * model); // sentence separator
     LLAMA_API llama_token llama_token_nl (const struct llama_model * model); // next-line
     LLAMA_API llama_token llama_token_pad(const struct llama_model * model); // padding
+    LLAMA_API llama_token llama_token_unk(const struct llama_model * model); // unknown
 
     LLAMA_API bool llama_add_bos_token(const struct llama_model * model);
     LLAMA_API bool llama_add_eos_token(const struct llama_model * model);
diff --git a/src/llama-vocab.cpp b/src/llama-vocab.cpp
index a4c01548..86da49bc 100644
--- a/src/llama-vocab.cpp
+++ b/src/llama-vocab.cpp
@@ -1681,6 +1681,10 @@ llama_token llama_token_pad_impl(const struct llama_vocab & vocab) {
     return vocab.special_pad_id;
 }
 
+llama_token llama_token_unk_impl(const struct llama_vocab & vocab) {
+    return vocab.special_unk_id;
+}
+
 bool llama_add_bos_token_impl(const struct llama_vocab & vocab) {
     return vocab.tokenizer_add_bos;
 }
diff --git a/src/llama-vocab.h b/src/llama-vocab.h
index 0d00086d..c413da42 100644
--- a/src/llama-vocab.h
+++ b/src/llama-vocab.h
@@ -128,6 +128,7 @@ llama_token llama_token_cls_impl(const struct llama_vocab & vocab);
 llama_token llama_token_sep_impl(const struct llama_vocab & vocab);
 llama_token llama_token_nl_impl (const struct llama_vocab & vocab);
 llama_token llama_token_pad_impl(const struct llama_vocab & vocab);
+llama_token llama_token_unk_impl(const struct llama_vocab & vocab);
 
 llama_token llama_token_prefix_impl(const struct llama_vocab & vocab);
 llama_token llama_token_middle_impl(const struct llama_vocab & vocab);
diff --git a/src/llama.cpp b/src/llama.cpp
index ae375bcd..15592891 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -12344,6 +12344,10 @@ llama_token llama_token_pad(const struct llama_model * model) {
     return llama_token_pad_impl(model->vocab);
 }
 
+llama_token llama_token_unk(const struct llama_model * model) {
+    return llama_token_unk_impl(model->vocab);
+}
+
 bool llama_add_bos_token(const struct llama_model * model) {
     return llama_add_bos_token_impl(model->vocab);
 }
