diff --git a/include/llama.h b/include/llama.h
index 168c3fa1..ed03d56a 100644
--- a/include/llama.h
+++ b/include/llama.h
@@ -902,6 +902,7 @@ extern "C" {
     LLAMA_API llama_token llama_token_sep(const struct llama_model * model); // sentence separator
     LLAMA_API llama_token llama_token_nl (const struct llama_model * model); // next-line
     LLAMA_API llama_token llama_token_pad(const struct llama_model * model); // padding
+    LLAMA_API llama_token llama_token_unk(const struct llama_model * model); // unknown
 
     LLAMA_API bool llama_add_bos_token(const struct llama_model * model);
     LLAMA_API bool llama_add_eos_token(const struct llama_model * model);
diff --git a/src/llama-vocab.cpp b/src/llama-vocab.cpp
index d1dc9627..2255ef95 100644
--- a/src/llama-vocab.cpp
+++ b/src/llama-vocab.cpp
@@ -1687,6 +1687,10 @@ llama_token llama_token_pad_impl(const struct llama_vocab & vocab) {
     return vocab.special_pad_id;
 }
 
+llama_token llama_token_unk_impl(const struct llama_vocab & vocab) {
+    return vocab.special_unk_id;
+}
+
 bool llama_add_bos_token_impl(const struct llama_vocab & vocab) {
     return vocab.tokenizer_add_bos;
 }
diff --git a/src/llama-vocab.h b/src/llama-vocab.h
index 4bb16d2e..528f0291 100644
--- a/src/llama-vocab.h
+++ b/src/llama-vocab.h
@@ -116,6 +116,7 @@ llama_token llama_token_cls_impl(const struct llama_vocab & vocab);
 llama_token llama_token_sep_impl(const struct llama_vocab & vocab);
 llama_token llama_token_nl_impl (const struct llama_vocab & vocab);
 llama_token llama_token_pad_impl(const struct llama_vocab & vocab);
+llama_token llama_token_unk_impl(const struct llama_vocab & vocab);
 
 llama_token llama_token_prefix_impl(const struct llama_vocab & vocab);
 llama_token llama_token_middle_impl(const struct llama_vocab & vocab);
diff --git a/src/llama.cpp b/src/llama.cpp
index 00f78639..05648091 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -21683,6 +21683,10 @@ llama_token llama_token_pad(const struct llama_model * model) {
     return llama_token_pad_impl(model->vocab);
 }
 
+llama_token llama_token_unk(const struct llama_model * model) {
+    return llama_token_unk_impl(model->vocab);
+}
+
 bool llama_add_bos_token(const struct llama_model * model) {
     return llama_add_bos_token_impl(model->vocab);
 }
