diff --git a/src/llama-model.cpp b/src/llama-model.cpp
index 7260cb15..47886fa0 100644
--- a/src/llama-model.cpp
+++ b/src/llama-model.cpp
@@ -1383,7 +1383,8 @@ void llm_load_vocab(llama_model_loader & ml, llama_model & model) {
                 tokenizer_pre == "megrez") {
                 vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_QWEN2;
             } else {
-                throw std::runtime_error(format("unknown pre-tokenizer type: '%s'", tokenizer_pre.c_str()));
+                LLAMA_LOG_WARN("%s: missing or unknown pre-tokenizer type, using: 'default'\n", __func__);
+                vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_DEFAULT;
             }
         } else if (vocab.type == LLAMA_VOCAB_TYPE_SPM) {
             vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_DEFAULT;
